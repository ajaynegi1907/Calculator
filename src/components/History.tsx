export interface HistoryRecord {
    expression: string;
    result: number;
}

export interface HistoryProps {
onClose: () => void;
setCurrentExpression: (expression: string) => void;
setResult: (value?: number) => void;
setHistoryData: (data: HistoryRecord[]) => void;
historyData: HistoryRecord[];
}

export const History = (props:HistoryProps) => {
    
    const onRecordClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>): void =>{
        const target = event.currentTarget;
        const expression = target.children[0].innerHTML;
        const result = parseInt(target.children[1].innerHTML);
        props.setCurrentExpression(expression);
        props.setResult(result);
    }

    const onClearAllClick = () => {
        props.setHistoryData([]);
    }

    const recordList = props.historyData.map( (item, i) => {
        return (
            <div key={i} className = "history-record" onClick= {onRecordClick}>
                <div className="expression">{item.expression}</div>
                <div className="result"> {item.result}</div>
            </div>
        )
    });
    return (
        <div className="history-container">
            <div className="history-close-btn" onClick={props.onClose}>X</div>
            <div className="history-body">{recordList}</div>
            <button className="clear-history-btn" onClick={onClearAllClick}>Clear all</button>
        </div>
    );
}