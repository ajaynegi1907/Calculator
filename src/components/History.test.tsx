import { shallow, ShallowWrapper } from "enzyme";
import { History, HistoryProps } from "./History";

const props = {
  onClose: jest.fn(),
  setCurrentExpression: jest.fn(),
  setResult: jest.fn(),
  setHistoryData: jest.fn(),
  historyData: [{ expression: "3*5", result: 8 }, { expression: "4/3", result: 1.33}],
};

let wrapper: ShallowWrapper;
beforeEach(() => {
  wrapper = shallow(<History {...props} />);
});

afterEach(() => {
  jest.clearAllMocks();
});

it("should match snapshot", () => {
  wrapper = shallow(<History {...props} />);
  expect(wrapper).toMatchSnapshot();
});

it("should close on cross click", () => {
    const closeButton = wrapper.find(".history-close-btn");
    closeButton.simulate("click");
    expect(props.onClose).toHaveBeenCalledTimes(1);
});

it("should set current expression and record with the clicked record value", () => {
    const clickedRecord = props.historyData[0];
    const firstRecord = wrapper.find(".history-body").childAt(0);
    const mockEvent = {
        currentTarget: {
            children:[
                {innerHTML: clickedRecord.expression},
                {innerHTML: clickedRecord.result}
            ]
        }
    }
    firstRecord.simulate("click", mockEvent);
    expect(props.setCurrentExpression).toHaveBeenCalled();
    expect(props.setResult).toHaveBeenCalled();
    let actualArgument = props.setCurrentExpression.mock.calls[0][0];
    expect(actualArgument).toEqual(clickedRecord.expression);
    actualArgument = props.setResult.mock.calls[0][0];
    expect(actualArgument).toEqual(clickedRecord.result);
});

it("should clear all records when clear all button is  clicke", ()=> {
    const clearAllButton  = wrapper.find(".clear-history-btn");
    clearAllButton.simulate("click");
    expect(props.setHistoryData).toHaveBeenCalledWith([]);
})

