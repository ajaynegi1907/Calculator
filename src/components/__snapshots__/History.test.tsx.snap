// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="history-container"
>
  <div
    className="history-close-btn"
    onClick={[MockFunction]}
  >
    X
  </div>
  <div
    className="history-body"
  >
    <div
      className="history-record"
      key="0"
      onClick={[Function]}
    >
      <div
        className="expression"
      >
        3*5
      </div>
      <div
        className="result"
      >
         
        8
      </div>
    </div>
    <div
      className="history-record"
      key="1"
      onClick={[Function]}
    >
      <div
        className="expression"
      >
        4/3
      </div>
      <div
        className="result"
      >
         
        1.33
      </div>
    </div>
  </div>
  <button
    className="clear-history-btn"
    onClick={[Function]}
  >
    Clear all
  </button>
</div>
`;
